### PR TITLE
Fixed Taxonomy Term Limit On Product Collection Filters

### LINKED ISSUE #46850: Product Collection: Product Categories Filter Only Supports First 10 Categories
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When you add a "Product categories" filter to a Product Collection block the UI will not necessarily let you select all of categories on your store.

### Expected behavior

You should be able to search for any category in your store and select it.

### Actual behavior

When taxonomy inspector controls are loaded they prefetch "all" of the terms for the taxonomy. This is done because the `FormTokenField` only accepts a list of string suggestions. We use this to map suggestions back to term IDs so that they can be assigned to the collection's query. Unfortunately, we're actually only selecting the first 10 terms. This means that even though terms will show up in the search they can't be selected since there is no ID to resolve to.

### Steps to reproduce

1. Ensure that you have more than 10 categories.
2. Create a Product Collection block.
3. Add a "Product categories" filter and attempt to add every category (just 11 or more unique ones)
4. You can double-check in the inspector and you'll see that only the categories in the initial `product_cat` query are usable. Any that aren't in this first request don't work.

### WordPress Environment

N/A

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by ObliviousHarmony**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This pull request refactors the `TaxonomyItem` component. Previously, `useEntityRecords` was used to populate a cache of all of the terms, but without any pagination, this list only contains the first 10 terms. We don't really _need_ this cache since the query for existing terms and search terms will provide all of the information we need to display suggestions and render tokens. With this in mind I have moved towards using `getEntityRecords` directly from the data store. I was also able to reduce the amount of code in the process.

Screenshot 2024-05-03 at 17 58 08

There are now three different presentations for term names:

* Terms that are missing (have been deleted but not removed from block) are notated as missing.
* Terms with duplicates have the ID appended to differentiate them.
* Terms will otherwise show the name.

Closes #46850.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add a "Product categories" and "Product tags" filter to a Product Collection block.
2. Perform the search and add a category. Confirm that there is no limit to the number of terms allowed on the site.
3. Create two categories with the same name and add them. Note that when they are both present they both have an ID.
4. Delete a term after saving it to the block. It should show that it's missing in the filter block on next view.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/changelog/fix-46850-collection-taxonomy-filter-pagination [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: fix
+
+Allow category and tag texonomy filters on the Product Collection block to see more than the first 10 categories.
```
Comment (Aljullu) [inline plugins/woocommerce/changelog/fix-46850-collection-taxonomy-filter-pagination]:
```suggestion
Allow category and tag taxonomy filters on the Product Collection block to see more than the first 10 categories.
```

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/taxonomy-controls/taxonomy-item.tsx [hunk 2]
```diff
@@ -20,253 +19,203 @@ interface TaxonomyItemProps {
...
-			if ( termId ) {
-				newTermIds.push( termId );
+	// Since our tokens include some encoding we need to perform some transformations
+	// before they can be displayed in the input and in the suggestion list.
+	const displayTermName = ( display: string ) => {
+		const term = getTermFromToken( display );
+		if ( term ) {
+			// Terms that are missing will be identified as such.
+			if ( ! term.name ) {
+				display = `(#${ term.id } Missing)`;
```
Comment (Aljullu) [inline plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/taxonomy-controls/taxonomy-item.tsx]: I guess this string should be localizable, no?

**Review by Aljullu — APPROVED** [review]
Nice improvement! I tested it and everything works as expected. The code looks good as well.

I left a couple of suggestions, but pre-approving anyway. I also think it would be a good idea to add automated tests to make sure the duplicate taxonomies and missing taxonomies are rendered with the correct labels. But I don't think it should block this PR as those tests didn't exist before anyway.

File: plugins/woocommerce/changelog/fix-46850-collection-taxonomy-filter-pagination (hunk 1 above)
Comment (ObliviousHarmony) [reply plugins/woocommerce/changelog/fix-46850-collection-taxonomy-filter-pagination]: Nice catch!

File: plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/taxonomy-controls/taxonomy-item.tsx (hunk 2 above)
Comment (ObliviousHarmony) [reply plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/taxonomy-controls/taxonomy-item.tsx]: You're right, thanks for the prompting. I don't think I've ever created a localized string before :smile:

**Pull request merged**